cpp/src/arrow/python/pandas_convert.cc

```diff
@@ -413,6 +418,52 @@ inline Status PandasConverter::ConvertData<BooleanType>(std::shared_ptr<Buffer>*
   return Status::OK();
 }
 
+Status InvalidConversion(PyObject* obj, const std::string& expected_type_name) {
+  PyObject* type = PyObject_Type(obj);
+
+  if (type == nullptr) {
+    DCHECK_NE(PyErr_Occurred(), nullptr)
+        << "Call to PyObject_Type(...) returned NULL but PyErr_Occurred() was NULL";
+    RETURN_IF_PYERROR();
+  }
+
+  PyObject* type_name = PyObject_GetAttrString(type, "__name__");
+
+  if (type_name == nullptr) {
+    Py_DECREF(type);
+    DCHECK_NE(PyErr_Occurred(), nullptr) << "Call to PyObject_GetAttrString(...) "
+                                            "returned NULL but PyErr_Occurred() was NULL";
+    RETURN_IF_PYERROR();
+  }
+
+  PyObject* bytes_obj = PyUnicode_AsUTF8String(type_name);
+
+  if (bytes_obj == nullptr) {
+    Py_DECREF(type_name);
+    Py_DECREF(type);
+    DCHECK_NE(PyErr_Occurred(), nullptr) << "Call to PyObject_AsUTF8String(...) returned "
+                                            "NULL but PyErr_Occurred() was NULL";
+    RETURN_IF_PYERROR();
+  }
+
+  Py_ssize_t size = PyBytes_GET_SIZE(bytes_obj);
+  const char* bytes = PyBytes_AS_STRING(bytes_obj);
+
+  DCHECK_NE(bytes, nullptr) << "bytes from type(...).__name__ were null";
+
+  std::string cpp_type_name(bytes, size);
+
+  std::stringstream ss;
+  ss << "Python object of type " << cpp_type_name << " is not None and is not a "
+     << expected_type_name << " object";
+  auto s = Status::TypeError(ss.str());
+
+  Py_DECREF(bytes_obj);
+  Py_DECREF(type_name);
+  Py_DECREF(type);
```

Comment:
You could try using the `OwnedRef` for these, save you a bunch of DECREFs per above